### PR TITLE
E2Ev2: Allow for extension debugger

### PR DESCRIPTION
**Pull request by RXminuS**
Before there was no good way of attaching a debugger to the Cody extension as it ran in the VSCode code-host. This made it difficult to debug outside of running the special VSCode "Run Extension" command.

Now we allow supplying a specific debugger port to the extension host that gets started as part of the test fixture. This then allows you to easily connect a debugger to the extension as part of the test.

This not only makes it easier to debug issues with flaky tests but also opens the door for faster iteration as you can now simply record a few UI interactions an play them in a loop as you edit and re-debug the code.

https://github.com/user-attachments/assets/1d7ff51e-ed19-4ede-9056-10774227880a

## Test Plan

No code outside of e2ev2 is affected. No additional testing required

### CODE REVIEW

**Comment by RXminuS** [comment]
* **#5003** <a href="https://app.graphite.dev/github/pr/sourcegraph/cody/5003?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#4992** <a href="https://app.graphite.dev/github/pr/sourcegraph/cody/4992?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#4993** <a href="https://app.graphite.dev/github/pr/sourcegraph/cody/4993?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* `main`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @RXminuS and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

File: vscode/e2e/utils/vscody/fixture.ts [hunk 1]
```diff
@@ -850,40 +880,116 @@ fixture.beforeAll(async () => {
...
+                    'vs',
+                    'server',
+                    'node',
+                    'server.main.js'
+                )
+            )
+            // We patch the server.main.js file to accept a environment variable
+            // (EXTENSION_HOST_INSPECT_ENV) that allows us to control the debug port
+            // of specifically the extension host. There is currently no offical
+            // mechanism to do so in the serve-web command.
```
Comment (keegancsmith) [inline vscode/e2e/utils/vscody/fixture.ts:915]: fancy, nice

File: vscode/e2e/utils/vscody/fixture.ts [hunk 2]
```diff
@@ -850,40 +880,116 @@ fixture.beforeAll(async () => {
...
+                    throw new Error(
+                        `Could not find variable name for insertion point for patching ${file}`
+                    )
+                }
+                //we split at the insertion point
+                const newContent =
+                    contents.slice(0, insertionPoint.index) +
+                    `process.env.EXTENSION_HOST_INSPECT_ENV ? ${varname}.execArgv.push(process.env.EXTENSION_HOST_INSPECT_ENV) : null;` +
+                    contents.slice(insertionPoint.index)
+                await fs.writeFile(file, newContent, 'utf-8')
```
Comment (keegancsmith) [inline vscode/e2e/utils/vscody/fixture.ts:948]: alternative, interested in your thoughts. Can't we just store a `.patch` file and then shell out to the patch binary?

File: vscode/e2e/utils/vscody/uix/vscode.ts [hunk 3]
```diff
@@ -40,12 +41,43 @@ export class Sidebar {
...
     executeCommand,
     workspaceDir,
 }: Pick<UIXContextFnContext, 'page' | 'vscodeUI' | 'executeCommand' | 'workspaceDir'>) {
+    // If we have a debugger then we show some user instructions
+    const { extensionHostDebugPort } = vscodeUI
+    if (extensionHostDebugPort) {
+        await t.step('Show Debug Instructions', async () => {
+            // load a landing page before the session is initialized
+            const placeholderDir = path.join(__dirname, '..', 'resources', 'vscode-placeholder')
+            await page.route('http://vscode-pending.local/**', async (route, request) => {
```
Comment (keegancsmith) [inline vscode/e2e/utils/vscody/uix/vscode.ts:56]: you include quite a lot of resources here? Does it need to be so fancy?

Comment (RXminuS) [reply vscode/e2e/utils/vscody/uix/vscode.ts:56]: I didn't want it to look like an error page in cases it was confusing.

**Pull request closed without merging**